### PR TITLE
feat: add typescript support for useStore<T>

### LINKED ISSUE #1112: Preserve or generate the return types of state methods
### I am interested in helping provide a fix!

Yes

### Which generators are impacted?

- [X] All
- [ ] Angular
- [ ] HTML
- [ ] Qwik
- [ ] React
- [ ] React-Native
- [ ] Solid
- [ ] Stencil
- [ ] Svelte
- [ ] Vue
- [ ] Web components

### Reproduction case

https://mitosis.builder.io/?outputTab=G4VwpkA%3D&code=PTAEAsBdIBwZwFwgOYEtLgK4CMB0BjAewFtgAhTVAGwBMBTAJwEkB5YY9QuVOYbKwtnYBDVADtgMYfgDWw5HV5EGdYHAb5gAfS2RFkODuA1hkYXwGylhKpmJi4uBsIDuuAwA8AUF9TEYhAyQoADeoJhwdADKkIF0ADSgAGKBoAC%2BoABmDCSgAOQAAtiUtIy4qITsnNxweQDcPpAAnjB0oADCNnZioAC8oV6goERieqMIoMJiTQ1DIKAAKiwAIiwTTP5UdMR0o6AYPKAu4LugcJgwAUFZqTRNYsIc%2BB1RUaBUUzRwg0eoNBgAfgmYjs2EYDTSDS8dA8V2C4j0DEy0janVs9gACjl4AMhkR0Q4gR0uvYANoAXShczAS1W602212wQOcCOJx650ugWCmVu90eqGe7Ve70%2B3yGcCk%2BDoRJBxDBDFmoHmtLWoA2MC2Oz2LLZp05cJuDFAdweTxebw%2BYi%2BPzgZlkaO6cAAgpAiXkzPw6JA8qAAD75YiCah0X0BvJiOgAN0Y9R%2BKpWao1WqZ%2B3Ah2O%2Bouht5xtNAqFIqtNqGKhjDEijvscAA6uyYtIZHQaETsIQbHQphCfDDDfRkZgqDzMGJ8JAKj0q2IABQwbGIYkErGEeAASlxw0IDmCdtMbX6EWisRU05CPyGCkgU7g09XEynFI3Q1L3swDB6c5XjnxTv9AYpSpDGk8TnqAl4AOKYNAjBRKgABedC3sCoKME%2Bz4qJAb49M0rSEJkoCfvAuCSiifS9P0EYoQwvoAgR87EVKbR%2BgGAAMoATAATCxgHpCBz5gd6NZ%2FBg07iPQHjIfKjDrme%2FF4ludqbgSrL9Luei4Je163jxL6Ye%2BoDTj%2B1akmJMLkqAABkFlKU6JnWmZuAuMJ4DrsxoAAIwsWxYBGQ4uBbGIyAYDxwGgZpJJiO0cC1s5on2RJoBygqMmgfJ242dWfRnGY6nhcp2mpZu6XIFBiKwQhWVqXQGnepB0EMOViGrjpRWKec2CQM4Y5Cf84BZdOJX1Y1oAAFQGb5jgBUFfUALQeau64%2BRFk27NNLUYVhoAAAb4MIVD4NOAAkIRVTVkA9SJpkeKuaQAKSgHNx3tZ10jnc5aQwNdW0hXx6TNT4umbdOoEADw0KgUaFfgHzRb0ABExTUPQDAzRNcNQ7DISyXJQzg5KHxNBMeSZFsHh5L9OMkzCyyoCoY4TkTvnk4VQx7agyBiEwejEAueR2hh%2BDgMzONDFQ4h0AAEnQ7NQETYiBMQe3CyLhQ7ODwgGYrHgzU5vUTAAnIbn2rnkEzYyLWSkzTdPjluRM5C4ys46FFuq82qAa9OWs685EwAGwAMz68bptoZTVu03Q9N2%2FkDszWWjCRE7ckuynaSgQAfIVIMpManaC70ISEd%2By1pFnFshIZEWJFd669BnBks8%2BYMQ8MMNwPDiOlCjvlw8MmNhFTHjgQ7RPub6aRl03%2FEhL5BBbmMkAZCEV3pxbQwg8YEPlyLN3Z8Auc7xvW9Rjv%2F3p0AA%3D

### Expected Behaviour

> What we currently preserve is the argument types, and all types inside of the body. We don't preserve or generate the return types of state methods yet.

raw file : https://github.com/BuilderIO/mitosis/blob/main/packages/core/src/__tests__/data/blocks/columns.raw.tsx#L25

```tsx
const state = useStore({
  getColumns(): Column[] {
    return props.columns || [];
  },
  getGutterSize(): number {
    return typeof props.space === 'number' ? props.space || 0 : 20;
  },
```

snapshot: https://github.com/BuilderIO/mitosis/pull/1085/files#diff-5c26547ec6bd2e4158291b871805f8465adcd4f2ffdd69227d4f028e643737a7L3770

```js
methods: {
  getColumns() { // here should has type Column[]
    return this.columns || [];
  },
  getGutterSize() { // here should has type number
    return typeof this.space === \\"number\\" ? this.space || 0 : 20;
  },
```

_Originally posted by @kingzez in https://github.com/BuilderIO/mitosis/issues/909#issuecomment-1503766963_
            

### Actual Behaviour

We don't preserve or generate the return types of state methods yet.

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by nmerget**
## Description

Please provide the following information:

Add TypeScript support for `useStore<T>` for those generators:

- [x] React
- [x] Angular
- [x] Vue

The implementation should work for other generators as well, but I'm not into most outputs. `typeParameter` is provided for `useStore` correctly, so it might work (Vue works out of the box, because of the correct implementation)

closes #1112

Make sure to follow the PR preparation steps in [CONTRIBUTING.md](../CONTRIBUTING.md#preparing-your-pr) before submitting your PR:

- [x] format the codebase: from the root, run `yarn fmt:prettier`.
- [x] update all snapshots (in core & CLI): from the root, run `yarn test:update`
- [x] add Changeset entry: from the root, run `yarn g:changeset` and follow the CLI instructions. Alternatively, use the Changeset Github Bot to create the file.

### CODE REVIEW

File: packages/core/src/generators/react/state.ts [hunk 1]
```diff
@@ -36,25 +37,26 @@ const processStateValue = (options: ToReactOptions) => {
...
-          return pipe(value, replaceGetterWithFunction, mapValue);
-        case 'function':
-          return mapValue(value);
-        case 'method':
-          return pipe(value, prefixWithFunction, mapValue);
-        default:
-          return getDefaultCase();
-      }
+    const typeParameter = stateVal.typeParameter;
+    const stateType = options.typescript ? `<${stateVal.typeParameter ?? 'any'}>` : '';
```
Comment (samijaber) [inline packages/core/src/generators/react/state.ts:44]: adding `any` here is actually causing more harm than good: TypeScript is always able to infer the type on its own, and you're overriding that:

CleanShot 2024-09-27 at 17 41 25@2x

if there is no `stateVal.typeParameter`, we should not do anything.

**Review by samijaber — CHANGES_REQUESTED** [review]
Awesome! I'm requesting changes to the React generator because the `any` default is bad, but otherwise LGTM 👍🏽

File: .changeset/quick-crabs-cough.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@builder.io/mitosis': patch
+---
+
+feat: add typescript support for useStore (angular,react,vue)
```
Comment (samijaber) [inline .changeset/quick-crabs-cough.md:5]: I just documented our changset format so we all write them the same way moving forward: https://github.com/BuilderIO/mitosis/pull/1571

https://github.com/BuilderIO/mitosis/blob/7ac77ab2ffeb59b9ce82e515f4466b933c89701e/CONTRIBUTING.md#changeset-format

File: packages/core/src/generators/react/state.ts (hunk 1 above)
Comment (nmerget) [reply packages/core/src/generators/react/state.ts]: Makes sense, I fixed it

File: .changeset/quick-crabs-cough.md (hunk 2 above)
Comment (nmerget) [reply .changeset/quick-crabs-cough.md]: Yes, I saw it thank you.

Fixed it.

File: packages/core/src/__tests__/test-generator.ts [hunk 3]
```diff
@@ -652,7 +675,13 @@ export const runTestsForTarget = <X extends BaseTranspilerOptions>({
                     expect(getOutput).toThrowError();
                   } else {
                     try {
-                      expect(getOutput()).toMatchSnapshot();
+                      const output = getOutput();
+                      if (logOutput) {
+                        process.stdout.write(`--- Start: ${key} ---\n\n`);
+                        process.stdout.write(output);
+                        process.stdout.write(`--- End: ${key} ---\n\n`);
+                      }
```
[the remark this answers is not in the document]
Comment (nmerget) [reply packages/core/src/__tests__/test-generator.ts:687]: At the moment I test the implementation by createing a temporary test like this (example for vue):

image

Then I look at `packages/core/src/__tests__/test-generator.ts` to find the name of the test case to put it into `only: [...]`.

How do you test the code, is there any other way right now?

[the remark this answers is not in the document]
Comment (samijaber) [reply packages/core/src/__tests__/test-generator.ts:687]: same as you, but i made a permanent file here: https://github.com/BuilderIO/mitosis/blob/main/packages/core/src/__tests__/local.test.ts

it's always skipped. I add the tests that i want into it and run `yarn test local`.

File: .changeset/quick-crabs-cough.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
+---
+'@builder.io/mitosis': patch
+---
+
+[Angular, React, Vue] feat: add typescript support for ``useStore`` 
+
+[All] Add new snapshot tests to check if TypeScript works for ``useStore``
```
Comment (samijaber) [inline .changeset/quick-crabs-cough.md:7]: i wouldn't include testing improvements in the release notes

Comment (samijaber) [reply .changeset/quick-crabs-cough.md:7]: (unless they're huge test infra additions)

Comment (samijaber) [inline .changeset/quick-crabs-cough.md:7]:
```suggestion
```

**Review by samijaber — APPROVED** [review]

**Pull request merged**